Update Makefile to automatically install needed dependencies

Ajedi32 opened the pull request:
Updates the Makefile to automatically install dependencies used by the project before attempting to build.

Dependant on #3, because it also included updates to the README mentioning dependencies that weren't documented.

Rothera commented:
Github seems to think this PR has 18 commits on it and a new readme despite having merged the previous one. Is there something you can do about that?

In any case I'm rather ambivalent about this. This is an issue of development setup, not build process. I don't really want things "managing" my virtualenvs or imposing particular structure on them and it doesn't strike me as the right thing to do to be creating them automatically anyway. Additionally, I personally use two venvs, not one; one for Python 3 and one for Python 3.

Ajedi32 commented:
Yeah, I think it just doesn't realize the master branch has changed. I'll rebase so the changes show up correctly.

And that's a good point about `virtualenv`. I guess I can see how it might make sense to separate the step of setting up your development environment from the build process itself. On the other hand, I like the idea of automating as much of the setup process as possible. I wonder if there's a way to get the best of both worlds here...

Carson-Shook commented:
I hope this isn't going to destroy my build process on OS X... If it does, I'll have to do something about that... Probably in the form of a flag or something.

Ajedi32 commented:
@Carson-Shook I doubt it. Why? Is there any part of your build process that isn't already in the Makefile? Try checking out this branch and running `make`. Assuming you already have `virtualenv` installed it should just work. (And if not, `sudo pip install virtualenv` should fix that.)

Carson-Shook commented:
Ah, okay. So it's a python thing. That's good then. I've been out all day, and I haven't had time to look into it much. Wasn't sure if this was going to be some other dev tool I'd need to hunt down. Carry on. 👍🏼

Ajedi32 commented:
TODO (note to self): 

- [ ] Revert the changes to the Makefile in 1ecc86f
- [ ] Update README to explain how to setup a development environment manually

Ajedi32 commented:
Okay, I removed the automated setup of virtualenv and added some explanation to the README instead. This is a much more minor change now, but it should be good to merge.

Ajedi32 commented:
@Rothera Can you take a look at this when you get a chance? I took your suggestion in your previous comment and removed all the virtualenv stuff, so now the changes here are very minor (but I still found them helpful when working on BPM).

Rothera commented on Makefile:
```diff
@@ -47,11 +47,14 @@ ADDON_DATA = \
 
-default: build/betterponymotes.xpi build/chrome.zip build/BPM.safariextension build/export.json.bz2
+.PHONY: default
+default: build/betterponymotes.xpi build/chrome.zip build/BPM.safariextension.zip build/export.json.bz2
```
Can @Carson-Shook say anything about this extension change? I'm fairly certain Safari has its own fancy file extension it prefers, so `.zip` may not be appropriate (same principle as `.xpi`).

Ajedi32 replied:
This isn't actually changing what the extension _is_, it's just tweaking the Makefile dependencies so that it only rebuilds the extension if the source files actually changed. The ZIP file is actually generated on [line 169](https://github.com/Rothera/bpm/blob/7599435331aad001b353e6bdefc41cdc09a6e2f8/Makefile#L169), which this PR does not modify.

Carson-Shook replied:
Why exactly is there a "zip" being added to the end of this? safariextenion is the proper extension in this case.

Ajedi32 replied:
Again, this is just updating the Makefile dependency to match the artifact built on [line 169](https://github.com/Rothera/bpm/blob/7599435331aad001b353e6bdefc41cdc09a6e2f8/Makefile#L169) (which is not something added by this PR, it was added all the way back in b4635b9bb50ee8e122d714b046e8aec4e367fcf3).

I have no idea why a zip file is being produced in the first place, but if that file is what's being built then this is the correct way to reflect that in the Makefile dependencies.

Carson-Shook replied:
I'm not sure why a zip is being made either. It might have been leftover from an early experiment that I was doing to try and skip packing the extension manually. I think it would be better to get rid of the zip entirely since it's unnecessary for how I'm currently building the extension.

Ajedi32 replied:
Sounds good. This PR seems like a decent place to do that ("Fixes and improvements to the build process") so I'll add that change to this PR tomorrow.

Rothera commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+pyyaml
+lxml
```
Is there a point to this? Otherwise it strikes me as clutter.

Ajedi32 replied:
These are both Python packages needed to build BPM. `requirements.txt` is a file used to list project dependencies for pip: https://pip.readthedocs.org/en/1.1/requirements.html
